# Preprocessing

First, let's get all of the imports out of the way:

In [1]:
#Dataframe manipulation library
import pandas as pd
#Math functions, we'll only need the sqrt function so let's import only that
from math import sqrt
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

Now let's read each file into their Dataframes:

In [2]:
#Storing the artists information into a pandas dataframe
artists_df = pd.read_csv('C:/Users/Naveena/Documents/Web Mining/Final project/hetrec2011-lastfm-2k/artists.csv')
#Storing the tags information into a pandas dataframe
tags_df = pd.read_csv('C:/Users/Naveena/Documents/Web Mining/Final project/hetrec2011-lastfm-2k/tags.csv')
#Storing the user artists information into a pandas dataframe
user_artists_df = pd.read_csv('C:/Users/Naveena/Documents/Web Mining/Final project/hetrec2011-lastfm-2k/user_artists.csv')
#Storing the user tagged artists information into a pandas dataframe
user_taggedartists_df = pd.read_csv('C:/Users/Naveena/Documents/Web Mining/Final project/hetrec2011-lastfm-2k/user_taggedartists.csv')

In [3]:
print('Shape of artists_df:',artists_df.shape)
print('Shape of tags_df:',tags_df.shape)
print('Shape of user_artists_df:',user_artists_df.shape)
print('Shape of user_taggedartists_df:',user_taggedartists_df.shape)

Shape of artists_df: (17632, 4)
Shape of tags_df: (11946, 2)
Shape of user_artists_df: (92834, 3)
Shape of user_taggedartists_df: (186479, 6)


In [4]:
artists_df.describe(include = 'all')[['id', 'name']]

id    name
count   17632.000000   17632
unique           NaN   17632
top              NaN  Occult
freq             NaN       1
mean     9156.636853     NaN
std      5392.515296     NaN
min         1.000000     NaN
25%      4501.750000     NaN
50%      9048.500000     NaN
75%     13769.250000     NaN
max     18745.000000     NaN

In [5]:
tags_df.describe(include = 'all')

tagID tagValue
count   11946.000000    11946
unique           NaN    11944
top              NaN   #NAME?
freq             NaN        3
mean     6242.315336      NaN
std      3667.498057      NaN
min         1.000000      NaN
25%      3036.250000      NaN
50%      6210.500000      NaN
75%      9460.750000      NaN
max     12648.000000      NaN

In [6]:
user_artists_df.describe(include = 'all')

userID      artistID        weight
count  92834.000000  92834.000000   92834.00000
mean    1037.010481   3331.123145     745.24393
std      610.870436   4383.590502    3751.32208
min        2.000000      1.000000       1.00000
25%      502.000000    436.000000     107.00000
50%     1029.000000   1246.000000     260.00000
75%     1568.000000   4350.000000     614.00000
max     2100.000000  18745.000000  352698.00000

In [7]:
user_taggedartists_df.describe(include = 'all')

userID       artistID          tagID            day  \
count  186479.000000  186479.000000  186479.000000  186479.000000   
mean     1035.600137    4375.845328    1439.582913       1.095566   
std       622.461272    4897.789595    2775.340279       0.712813   
min         2.000000       1.000000       1.000000       1.000000   
25%       488.000000     686.000000      79.000000       1.000000   
50%      1021.000000    2203.000000     195.000000       1.000000   
75%      1624.000000    6714.000000     887.000000       1.000000   
max      2100.000000   18744.000000   12647.000000       9.000000   

               month           year  
count  186479.000000  186479.000000  
mean        6.524215    2008.809791  
std         3.486855       1.410062  
min         1.000000    1956.000000  
25%         3.000000    2008.000000  
50%         7.000000    2009.000000  
75%        10.000000    2010.000000  
max        12.000000    2011.000000

### Merging columns

<p>Merge columns (artistid, artistname, tagid) and store them in a new dataframe artists.</p>

In [8]:
#rename the column id in artists_df to artistID to make easier to join the dataframes.
artists_df.rename(columns = {'id': 'artistID'}, inplace = True)

#initially merge artists_df and user_taggedartists_df as they both contain artistID
artists = pd.merge(artists_df, user_taggedartists_df, on='artistID', how='inner')

#Then merge artists and tags_df
artists = pd.merge(artists, tags_df, on = 'tagID', how = 'inner')
print(artists.shape)
artists.head()

(184941, 10)


artistID                      name  \
0         1              MALICE MIZER   
1         8               DIR EN GREY   
2       374        宇多田ヒカル   
3      1929  ASIAN KUNG-FU GENERATION   
4         1              MALICE MIZER   

                                                 url  \
0              http://www.last.fm/music/MALICE+MIZER   
1               http://www.last.fm/music/DIR+EN+GREY   
2  http://www.last.fm/music/%E5%AE%87%E5%A4%9A%E7...   
3  http://www.last.fm/music/ASIAN+KUNG-FU+GENERATION   
4              http://www.last.fm/music/MALICE+MIZER   

                                          pictureURL  userID  tagID  day  \
0    http://userserve-ak.last.fm/serve/252/10808.jpg      37    552    1   
1  http://userserve-ak.last.fm/serve/252/46968835...      37    552    1   
2  http://userserve-ak.last.fm/serve/252/55544057...      37    552    1   
3   http://userserve-ak.last.fm/serve/252/358164.jpg      37    552    1   
4    http://userserve-ak.last.fm/serve/252/10808.jpg     274   1219    1   

   month  year tagValue  
0      7  2008   weeabo  
1      7  2008   weeabo  
2      7  2008   weeabo  
3      7  2008   weeabo  
4      5  2008    jrock

Drop unnecessary columns.

In [9]:
artists.drop(['url', 'pictureURL', 'userID', 'day', 'month', 'year', 'tagValue'], inplace = True, axis = 1)
artists.head(10)

artistID                      name  tagID
0         1              MALICE MIZER    552
1         8               DIR EN GREY    552
2       374        宇多田ヒカル    552
3      1929  ASIAN KUNG-FU GENERATION    552
4         1              MALICE MIZER   1219
5         8               DIR EN GREY   1219
6         8               DIR EN GREY   1219
7         8               DIR EN GREY   1219
8       376                 UVERworld   1219
9       376                 UVERworld   1219

Since the name column contains non-alphanumeric values, initally replace those values with NaN and then drop rows that contain NaN values.

In [10]:
artists['name'].replace(r'\W', np.nan, regex=True, inplace=True)
artists.dropna(inplace = True)

In [11]:
artists.head(10)

artistID       name  tagID
8        376  UVERworld   1219
9        376  UVERworld   1219
10       390      GACKT   1219
13      1857        YUI   1219
14      1872       Flow   1219
23      4980     Olivia   1219
25      5077       hide   1219
33      8396      12012   1219
34      8397     VIDOLL   1219
35      8404      Sadie   1219

In [12]:
artists.reset_index(drop = True)

artistID          name  tagID
0           376     UVERworld   1219
1           376     UVERworld   1219
2           390         GACKT   1219
3          1857           YUI   1219
4          1872          Flow   1219
5          4980        Olivia   1219
6          5077          hide   1219
7          8396         12012   1219
8          8397        VIDOLL   1219
9          8404         Sadie   1219
10        12801   Megaromania   1219
11        12803         ViViD   1219
12        17719           Miz   1219
13          376     UVERworld    139
14          376     UVERworld    139
15          376     UVERworld    139
16          376     UVERworld    139
17          376     UVERworld    139
18          390         GACKT    139
19          390         GACKT    139
20          390         GACKT    139
21          390         GACKT    139
22          406          HYDE    139
23          406          HYDE    139
24         1857           YUI    139
25         1857           YUI    139
26         1857           YUI    139
27         1872          Flow    139
28         1872          Flow    139
29         1872          Flow    139
...         ...           ...    ...
54810     18232  THEreminGIRL  12232
54811     18229         VFSix  12233
54812     18232  THEreminGIRL  12233
54813     18229         VFSix  12234
54814     18232  THEreminGIRL  12234
54815     18229         VFSix  12235
54816     18232  THEreminGIRL  12235
54817     18229         VFSix  12236
54818     18232  THEreminGIRL  12236
54819     18283        Expose   5490
54820     18320          Lava   4747
54821     18341     Castanets   7705
54822     18424      Amycanbe   2633
54823     18424      Amycanbe  10818
54824     18424      Amycanbe  12455
54825     18424      Amycanbe  12456
54826     18424      Amycanbe  12457
54827     18424      Amycanbe  12458
54828     18424      Amycanbe  12459
54829     18424      Amycanbe  12460
54830     18424      Amycanbe  12461
54831     18615         YEFFI  12616
54832     18615         YEFFI  12617
54833     18615         YEFFI  12618
54834     18630  Doppelganger  12633
54835     18736     LOSTFREEQ   3661
54836     18736     LOSTFREEQ   3665
54837     18736     LOSTFREEQ   3666
54838     18736     LOSTFREEQ   3667
54839     18736     LOSTFREEQ   3672

[54840 rows x 3 columns]

In [13]:
artists.describe(include = 'all')

artistID     name         tagID
count   54840.000000    54840  54840.000000
unique           NaN     3407           NaN
top              NaN  Madonna           NaN
freq             NaN      730           NaN
mean     4157.318071      NaN   1345.595678
std      4756.374220      NaN   2713.337728
min         6.000000      NaN      1.000000
25%       707.000000      NaN     73.000000
50%      2094.000000      NaN    187.000000
75%      6160.000000      NaN    824.000000
max     18736.000000      NaN  12647.000000

As there are 54,840 records, which is quite a lot, let's take sample of 15% of the data for further analysis.

In [14]:
artists_sample = artists.sample(frac = 0.15)

#Let's check if artists_sample contains 15% of the artists data
if(0.15 * (len(artists)) == len(artists_sample)):
    print('artists_sample data frame contains 15% of artists data!')
    
print(artists_sample.shape)
artists_sample.describe(include = 'all')

artists_sample data frame contains 15% of artists data!
(8226, 3)


artistID     name         tagID
count    8226.000000     8226   8226.000000
unique           NaN     2121           NaN
top              NaN  Madonna           NaN
freq             NaN      102           NaN
mean     4154.797107      NaN   1379.472526
std      4730.780569      NaN   2751.212234
min         6.000000      NaN      1.000000
25%       702.500000      NaN     73.000000
50%      2160.500000      NaN    187.500000
75%      6130.000000      NaN    829.000000
max     18736.000000      NaN  12642.000000

Let's convert tagID into a string, to make it easier for creating dummy variables for the column.

In [15]:
artists_sample['tagID'] = artists_sample['tagID'].astype('str')
print(artists_sample.dtypes)

artists_sample.describe(include = 'all')

artistID     int64
name        object
tagID       object
dtype: object


artistID     name tagID
count    8226.000000     8226  8226
unique           NaN     2121  1610
top              NaN  Madonna    73
freq             NaN      102   356
mean     4154.797107      NaN   NaN
std      4730.780569      NaN   NaN
min         6.000000      NaN   NaN
25%       702.500000      NaN   NaN
50%      2160.500000      NaN   NaN
75%      6130.000000      NaN   NaN
max     18736.000000      NaN   NaN

<p>Since keeping tags in a list format isn't optimal for the content-based recommendation system technique, we will use the One Hot Encoding technique to convert the list of tag values to a vector where each column corresponds to one possible value of the feature.</p>

<p>This encoding is needed for feeding categorical data. In this case, we store every different tag value in columns that contain either 1 or 0. 1 shows that an artist has that genre and 0 shows that it doesn't.</p>

<p>Let's also store this dataframe in another variable since tags won't be important for our first recommendation system.</p>

In [16]:
#Copying the movie dataframe into a new one since we won't need to use the genre information in our first case.
artistsWithTags_df = artists_sample.copy()

artistsWithTags_df = pd.get_dummies(data = artists_sample, columns = ['tagID'])
print(artistsWithTags_df.shape)

artistsWithTags_df.head()

(8226, 1612)


artistID          name  tagID_1  tagID_10  tagID_1001  tagID_1006  \
4297        616  Squarepusher        0         0           0           0   
41364       173       Placebo        0         0           0           0   
73394     14225          Doom        0         0           0           0   
97758       965          Bush        0         0           0           0   
51029     14293         Blank        0         0           0           0   

       tagID_10092  tagID_10093  tagID_10096  tagID_10110  ...  tagID_9919  \
4297             0            0            0            0  ...           0   
41364            0            0            0            0  ...           0   
73394            0            0            0            0  ...           0   
97758            0            0            0            0  ...           0   
51029            0            0            0            0  ...           0   

       tagID_9928  tagID_9952  tagID_9954  tagID_9955  tagID_996  tagID_9969  \
4297            0           0           0           0          0           0   
41364           0           0           0           0          0           0   
73394           0           0           0           0          0           0   
97758           0           0           0           0          0           0   
51029           0           0           0           0          0           0   

       tagID_9972  tagID_9973  tagID_9997  
4297            0           0           0  
41364           0           0           0  
73394           0           0           0  
97758           0           0           0  
51029           0           0           0  

[5 rows x 1612 columns]

In [17]:
artistsWithTags_df.reset_index(drop = True)

artistID          name  tagID_1  tagID_10  tagID_1001  tagID_1006  \
0          616  Squarepusher        0         0           0           0   
1          173       Placebo        0         0           0           0   
2        14225          Doom        0         0           0           0   
3          965          Bush        0         0           0           0   
4        14293         Blank        0         0           0           0   
5         3540     Discharge        0         0           0           0   
6         6046       Nujabes        0         0           0           0   
7         5575         SS501        0         0           0           0   
8        11475            Ef        0         0           0           0   
9           32      Covenant        0         0           0           0   
10        4283    Sarcófago        1         0           0           0   
11         454        Jónsi        0         0           0           0   
12         877    Alphaville        0         0           0           0   
13        1787         UNKLE        0         0           0           0   
14         498      Paramore        0         0           0           0   
15        2346     Aerosmith        0         0           0           0   
16         726  Apocalyptica        0         0           0           0   
17        3105        Guster        0         0           0           0   
18       15649      Feldberg        0         0           0           0   
19         511            U2        0         0           0           0   
20       15880       Swayzak        0         0           0           0   
21         190          Muse        0         0           0           0   
22        2346     Aerosmith        0         0           0           0   
23        4290    Dissection        0         0           0           0   
24         154     Radiohead        0         0           0           0   
25       12035           CW7        0         0           0           0   
26        4395    Lungbutter        0         0           0           0   
27       14292          Numb        0         0           0           0   
28         946     Avantasia        0         0           0           0   
29        1096       Bayside        0         0           0           0   
...        ...           ...      ...       ...         ...         ...   
8196       493  Lostprophets        0         0           0           0   
8197        65      Coldplay        0         0           0           0   
8198      6887            AZ        0         0           0           0   
8199      1243          Mika        0         0           0           0   
8200      3637      Ektomorf        0         0           0           0   
8201      4406     Pixieguts        0         0           0           0   
8202       288       Rihanna        0         0           0           0   
8203      3918         Fancy        0         0           0           0   
8204      3490         Utada        0         0           0           0   
8205      8360       Ximbica        0         0           0           0   
8206     17320       Autolux        0         0           0           0   
8207      3983    Rosenstolz        0         0           0           0   
8208      3470       Rainbow        0         0           0           0   
8209     12583   iLiKETRAiNS        0         0           0           0   
8210      4728       Vitalic        0         0           0           0   
8211       288       Rihanna        0         0           0           0   
8212       511            U2        0         0           0           0   
8213        67       Madonna        0         0           0           0   
8214      2548       Wanessa        0         0           0           0   
8215     10859          Orgy        0         0           0           0   
8216       172       Garbage        0         0           0           0   
8217      8789        Darude   

Now, let's take a look at user_artists_df dataframe.

In [18]:
print(user_artists_df.shape)
user_artists_df.head()

(92834, 3)


userID  artistID  weight
0       2        51   13883
1       2        52   11690
2       2        53   11351
3       2        54   10300
4       2        55    8983

Every user is associated with an artist and number of times the user listened to the artist is represented as weight.

In [19]:
#Let's see on an average how many times each artists was listened by total number of users.
user_artists_df.groupby('artistID').agg({'userID':'count','weight':'mean'}).sort_values(['weight'], ascending = False)

userID        weight
artistID                      
8388           1  35323.000000
6373           1  30614.000000
18121          1  23462.000000
8308           2  16497.500000
14986          1  15345.000000
14987          1  14082.000000
792           26  13462.884615
15075          1  12520.000000
4625           1  10776.000000
18122          1  10726.000000
8366           1  10574.000000
9665           1   9975.000000
8281           1   9615.000000
15347          1   9453.000000
8367           1   9097.000000
4626           1   8833.000000
4627           1   8636.000000
14185          2   8615.500000
6696           1   8511.000000
15927          1   8320.000000
18123          1   8131.000000
5391           2   7976.000000
13182          2   7667.500000
10871          1   7666.000000
12356          1   7566.000000
17935          1   7500.000000
18205          1   7494.000000
12384          1   7423.000000
17936          1   7342.000000
2734           2   7101.500000
...          ...           ...
5921           1      1.000000
5920           1      1.000000
14375          1      1.000000
14376          1      1.000000
17172          1      1.000000
13712          1      1.000000
17171          1      1.000000
17170          1      1.000000
17169          1      1.000000
18433          1      1.000000
18432          1      1.000000
13825          1      1.000000
13824          1      1.000000
7612           1      1.000000
15749          1      1.000000
11644          1      1.000000
9481           1      1.000000
11248          1      1.000000
11250          1      1.000000
11251          1      1.000000
14373          1      1.000000
6261           1      1.000000
7752           1      1.000000
17080          1      1.000000
13823          1      1.000000
13762          1      1.000000
13761          1      1.000000
16413          1      1.000000
9482           1      1.000000
4552           1      1.000000

[17632 rows x 2 columns]

In [20]:
#Let's see on an average how many artists were listened by each user.
user_artists_df.groupby('userID').agg({'artistID':'mean'}).sort_values(['artistID'], ascending = False)

artistID
userID              
2085    18615.000000
1893    17466.500000
1758    16497.000000
1731    16364.000000
1847    15935.880000
1854    15900.720000
1603    15529.000000
1882    15083.580000
1989    14982.080000
2095    14763.780000
1499    14559.100000
1721    14117.120000
1930    13971.100000
1266    13564.000000
1719    13293.380000
1477    13244.660000
2005    13051.500000
1939    12857.875000
1867    12653.340000
1315    12627.220000
2016    12314.140000
1497    11852.460000
1090    11634.240000
1121    11570.540000
1205    11502.785714
1848    11498.340000
1777    11355.460000
1426    11205.180000
1221    11089.920000
1971    11046.860000
...              ...
1272      582.200000
24        581.460000
1808      580.420000
722       576.400000
14        573.500000
18        572.740000
965       564.960000
1606      551.080000
797       545.090909
425       540.040000
21        538.620000
1666      532.680000
17        527.840000
15        501.600000
25        478.740000
12        452.560000
13        421.980000
11        393.800000
10        383.980000
9         383.340000
8         317.300000
1773      314.600000
7         295.020000
701       279.272727
6         259.280000
5         209.440000
4         156.960000
3         125.500000
1307       89.000000
2          75.500000

[1892 rows x 1 columns]

# Content-Based recommendation system

Now, let's take a look at how to implement __Content-Based__ or __Item-Item recommendation systems__. This technique attempts to figure out what a user's favourite aspects of an item is, and then recommends items that present those aspects. In our case, we're going to try to figure out the input's favorite tags from the artists and ratings given.

Let's begin by creating an input user to recommend artists to:

Notice: To add more artists, simply increase the amount of elements in the __userInput__. Feel free to add more in!

In [21]:
userInput = [
            {'name':'OneRepublic', 'weight':4576},
            {'name':'MGMT', 'weight':484},
            {'name':'Madonna', 'weight':8473},
            {'name':"Shakira", 'weight':5875},
            {'name':'Rihanna', 'weight':4750},
            {'name':'Adele', 'weight':8887}
         ] 
inputArtists = pd.DataFrame(userInput)
inputArtists

name  weight
0  OneRepublic    4576
1         MGMT     484
2      Madonna    8473
3      Shakira    5875
4      Rihanna    4750
5        Adele    8887

#### Add artistId to input user
With the input complete, let's extract the input artist's ID's from the artists dataframe and add them into it.

We can achieve this by first filtering out the rows that contain the input artist's name and then merging this subset with the input dataframe. We also drop unnecessary columns for the input to save memory space.

In [22]:
#Filtering out the artists by name
inputId = artists_df[artists_df['name'].isin(inputArtists['name'].tolist())]

#Then merging it so we can get the artistId. It's implicitly merging it by name.
inputArtists = pd.merge(inputId, inputArtists)

#Dropping information we won't use from the input dataframe
inputArtists = inputArtists.drop('url', 1).drop('pictureURL', 1)
#Final input dataframe
inputArtists

artistID         name  weight
0        67      Madonna    8473
1       191  OneRepublic    4576
2       288      Rihanna    4750
3       701      Shakira    5875
4      1400         MGMT     484
5      1934        Adele    8887

We're going to start by learning the input's preferences, so let's get the subset of artists that the input has listened to from the Dataframe containing tags defined with binary values.

In [23]:
#Filtering out the artists from the input
userArtists = artistsWithTags_df[artistsWithTags_df['artistID'].isin(inputArtists['artistID'].tolist())]
userArtists

artistID         name  tagID_1  tagID_10  tagID_1001  tagID_1006  \
127835      1934        Adele        0         0           0           0   
23224        288      Rihanna        0         0           0           0   
156611       288      Rihanna        0         0           0           0   
77622        288      Rihanna        0         0           0           0   
166411      1400         MGMT        0         0           0           0   
60274        288      Rihanna        0         0           0           0   
41249         67      Madonna        0         0           0           0   
145625       701      Shakira        0         0           0           0   
76086       1934        Adele        0         0           0           0   
112348        67      Madonna        0         0           0           0   
79893         67      Madonna        0         0           0           0   
111420       288      Rihanna        0         0           0           0   
80482        288      Rihanna        0         0           0           0   
69336        288      Rihanna        0         0           0           0   
6161          67      Madonna        0         0           0           0   
146710      1400         MGMT        0         0           0           0   
109184      1934        Adele        0         0           0           0   
44490        288      Rihanna        0         0           0           0   
92553        701      Shakira        0         0           0           0   
57164         67      Madonna        0         0           0           0   
60290        288      Rihanna        0         0           0           0   
61039        701      Shakira        0         0           0           0   
60301        288      Rihanna        0         0           0           0   
79470         67      Madonna        0         0           0           0   
20966         67      Madonna        0         0           0           0   
102785       701      Shakira        0         0           0           0   
49966         67      Madonna        0         0           0           0   
129440       288      Rihanna        0         0           0           0   
59991         67      Madonna        0         0           0           0   
171448       701      Shakira        0         0           0           0   
...          ...          ...      ...       ...         ...         ...   
128946        67      Madonna        0         0           0           0   
113963       288      Rihanna        0         0           0           0   
80503        288      Rihanna        0         0           0           0   
6140          67      Madonna        0         0           0           0   
62782       1400         MGMT        0         0           0           0   
51274        288      Rihanna        0         0           0           0   
116700      1934        Adele        0         0           0           0   
156470       288      Rihanna        0         0           0           0   
160730       288      Rihanna        0         0           0           0   
114009       288      Rihanna        0         0           0           0   
144354       701      Shakira        0         0           0           0   
80511        288      Rihanna        0         0           0           0   
113222        67      Madonna        0         0           0           0   
130628        67      Madonna        0         0           0           0   
110784      1400         MGMT        0         0           0           0   
96430        701      Shakira        0         0           0           0   
96037        191  OneRepublic        0         0           0           0   
160675       288      Rihanna        0         0           0           0   
160495       288      Rihanna        0         0           0           0   
125764       191  OneRepublic        0         0           0           0   
143976       191  OneRepublic        0         0        

In [24]:
userArtists = userArtists.groupby(['artistID', 'name'], as_index = False).mean()
userArtists

artistID         name  tagID_1  tagID_10  tagID_1001  tagID_1006  \
0        67      Madonna      0.0       0.0         0.0         0.0   
1       191  OneRepublic      0.0       0.0         0.0         0.0   
2       288      Rihanna      0.0       0.0         0.0         0.0   
3       701      Shakira      0.0       0.0         0.0         0.0   
4      1400         MGMT      0.0       0.0         0.0         0.0   
5      1934        Adele      0.0       0.0         0.0         0.0   

   tagID_10092  tagID_10093  tagID_10096  tagID_10110  ...  tagID_9919  \
0          0.0          0.0          0.0          0.0  ...         0.0   
1          0.0          0.0          0.0          0.0  ...         0.0   
2          0.0          0.0          0.0          0.0  ...         0.0   
3          0.0          0.0          0.0          0.0  ...         0.0   
4          0.0          0.0          0.0          0.0  ...         0.0   
5          0.0          0.0          0.0          0.0  ...         0.0   

   tagID_9928  tagID_9952  tagID_9954  tagID_9955  tagID_996  tagID_9969  \
0         0.0         0.0         0.0         0.0        0.0         0.0   
1         0.0         0.0         0.0         0.0        0.0         0.0   
2         0.0         0.0         0.0         0.0        0.0         0.0   
3         0.0         0.0         0.0         0.0        0.0         0.0   
4         0.0         0.0         0.0         0.0        0.0         0.0   
5         0.0         0.0         0.0         0.0        0.0         0.0   

   tagID_9972  tagID_9973  tagID_9997  
0         0.0         0.0         0.0  
1         0.0         0.0         0.0  
2         0.0         0.0         0.0  
3         0.0         0.0         0.0  
4         0.0         0.0         0.0  
5         0.0         0.0         0.0  

[6 rows x 1612 columns]

We'll only need the actual tags table, so let's clean this up a bit by resetting the index and dropping the artistID and name column.

In [25]:
#Resetting the index to avoid future issues
userArtists = userArtists.reset_index(drop=True)

#Dropping unnecessary issues due to save memory and to avoid issues
userTagTable = userArtists.drop('artistID', 1).drop('name', 1)
userTagTable

tagID_1  tagID_10  tagID_1001  tagID_1006  tagID_10092  tagID_10093  \
0      0.0       0.0         0.0         0.0          0.0          0.0   
1      0.0       0.0         0.0         0.0          0.0          0.0   
2      0.0       0.0         0.0         0.0          0.0          0.0   
3      0.0       0.0         0.0         0.0          0.0          0.0   
4      0.0       0.0         0.0         0.0          0.0          0.0   
5      0.0       0.0         0.0         0.0          0.0          0.0   

   tagID_10096  tagID_10110  tagID_10111  tagID_10112  ...  tagID_9919  \
0          0.0          0.0          0.0          0.0  ...         0.0   
1          0.0          0.0          0.0          0.0  ...         0.0   
2          0.0          0.0          0.0          0.0  ...         0.0   
3          0.0          0.0          0.0          0.0  ...         0.0   
4          0.0          0.0          0.0          0.0  ...         0.0   
5          0.0          0.0          0.0          0.0  ...         0.0   

   tagID_9928  tagID_9952  tagID_9954  tagID_9955  tagID_996  tagID_9969  \
0         0.0         0.0         0.0         0.0        0.0         0.0   
1         0.0         0.0         0.0         0.0        0.0         0.0   
2         0.0         0.0         0.0         0.0        0.0         0.0   
3         0.0         0.0         0.0         0.0        0.0         0.0   
4         0.0         0.0         0.0         0.0        0.0         0.0   
5         0.0         0.0         0.0         0.0        0.0         0.0   

   tagID_9972  tagID_9973  tagID_9997  
0         0.0         0.0         0.0  
1         0.0         0.0         0.0  
2         0.0         0.0         0.0  
3         0.0         0.0         0.0  
4         0.0         0.0         0.0  
5         0.0         0.0         0.0  

[6 rows x 1610 columns]

Now we're ready to start learning the input's preferences!

To do this, we're going to turn each tag into weights. We can do this by using the input's weights and multiplying them into the input's tag table and then summing up the resulting table by column. This operation is actually a dot product between a matrix and a vector, so we can simply accomplish by calling Pandas's "dot" function.

In [26]:
inputArtists['weight']

0    8473
1    4576
2    4750
3    5875
4     484
5    8887
Name: weight, dtype: int64

In [27]:
#Dot produt to get weights
userProfile = userTagTable.transpose().dot(inputArtists['weight'])
#The user profile
userProfile

tagID_1          0.000000
tagID_10         0.000000
tagID_1001       0.000000
tagID_1006       0.000000
tagID_10092      0.000000
tagID_10093      0.000000
tagID_10096      0.000000
tagID_10110      0.000000
tagID_10111      0.000000
tagID_10112      0.000000
tagID_10116      0.000000
tagID_10133      0.000000
tagID_10138      0.000000
tagID_1016       0.000000
tagID_10162      0.000000
tagID_1017       0.000000
tagID_102      418.362745
tagID_10220      0.000000
tagID_1023       0.000000
tagID_10230      0.000000
tagID_10234      0.000000
tagID_10235      0.000000
tagID_1024       0.000000
tagID_10243      0.000000
tagID_1025       0.000000
tagID_10259      0.000000
tagID_1026       0.000000
tagID_10261      0.000000
tagID_1028       0.000000
tagID_10286     83.068627
                  ...    
tagID_9785       0.000000
tagID_9793       0.000000
tagID_9805       0.000000
tagID_9809       0.000000
tagID_9817       0.000000
tagID_9819       0.000000
tagID_9820       0.000000
tagID_9821  

Now, we have the weights for every of the user's preferences. This is known as the User Profile. Using this, we can recommend artists that satisfy the user's preferences.

Let's start by extracting the tag table from the original dataframe:

In [28]:
#Now let's get the genres of every movie in our original dataframe
tagTable = artistsWithTags_df.set_index(artistsWithTags_df['artistID'])

tagTable.head()

artistID          name  tagID_1  tagID_10  tagID_1001  tagID_1006  \
artistID                                                                      
616            616  Squarepusher        0         0           0           0   
173            173       Placebo        0         0           0           0   
14225        14225          Doom        0         0           0           0   
965            965          Bush        0         0           0           0   
14293        14293         Blank        0         0           0           0   

          tagID_10092  tagID_10093  tagID_10096  tagID_10110  ...  tagID_9919  \
artistID                                                      ...               
616                 0            0            0            0  ...           0   
173                 0            0            0            0  ...           0   
14225               0            0            0            0  ...           0   
965                 0            0            0            0  ...           0   
14293               0            0            0            0  ...           0   

          tagID_9928  tagID_9952  tagID_9954  tagID_9955  tagID_996  \
artistID                                                              
616                0           0           0           0          0   
173                0           0           0           0          0   
14225              0           0           0           0          0   
965                0           0           0           0          0   
14293              0           0           0           0          0   

          tagID_9969  tagID_9972  tagID_9973  tagID_9997  
artistID                                                  
616                0           0           0           0  
173                0           0           0           0  
14225              0           0           0           0  
965                0           0           0           0  
14293              0           0           0           0  

[5 rows x 1612 columns]

In [29]:
#And drop the unnecessary information
tagTable = tagTable.drop('artistID', 1).drop('name', 1)
tagTable.head()

tagID_1  tagID_10  tagID_1001  tagID_1006  tagID_10092  tagID_10093  \
artistID                                                                        
616             0         0           0           0            0            0   
173             0         0           0           0            0            0   
14225           0         0           0           0            0            0   
965             0         0           0           0            0            0   
14293           0         0           0           0            0            0   

          tagID_10096  tagID_10110  tagID_10111  tagID_10112  ...  tagID_9919  \
artistID                                                      ...               
616                 0            0            0            0  ...           0   
173                 0            0            0            0  ...           0   
14225               0            0            0            0  ...           0   
965                 0            0            0            0  ...           0   
14293               0            0            0            0  ...           0   

          tagID_9928  tagID_9952  tagID_9954  tagID_9955  tagID_996  \
artistID                                                              
616                0           0           0           0          0   
173                0           0           0           0          0   
14225              0           0           0           0          0   
965                0           0           0           0          0   
14293              0           0           0           0          0   

          tagID_9969  tagID_9972  tagID_9973  tagID_9997  
artistID                                                  
616                0           0           0           0  
173                0           0           0           0  
14225              0           0           0           0  
965                0           0           0           0  
14293              0           0           0           0  

[5 rows x 1610 columns]

In [30]:
tagTable.shape

(8226, 1610)

With the input's profile and the complete list of artists and their tags in hand, we're going to take the weighted average of every artist based on the input profile and recommend the top twenty artists that most satisfy it.

In [31]:
#Multiply the genres by the weights and then take the weighted average
recommendationTable_df = ((tagTable*userProfile).sum(axis=1))/(userProfile.sum())
recommendationTable_df.head()

artistID
616      0.000000
173      0.017597
14225    0.000000
965      0.000000
14293    0.002514
dtype: float64

In [32]:
#Sort our recommendations in descending order
recommendationTable_df = recommendationTable_df.sort_values(ascending=False)
#Just a peek at the values
recommendationTable_df.head()

artistID
4704     0.059801
288      0.059801
4609     0.059801
701      0.059801
15171    0.059801
dtype: float64

Now here's the recommendation table!

In [33]:
#The final recommendation table
final_recomm = artists_df.loc[artists_df['artistID'].isin(recommendationTable_df.head(20).keys())]

In [34]:
final_recomm

artistID        name                                  url  \
65           71        Dido        http://www.last.fm/music/Dido   
282         288     Rihanna     http://www.last.fm/music/Rihanna   
314         320      Fergie      http://www.last.fm/music/Fergie   
334         340        Cher        http://www.last.fm/music/Cher   
456         462        Blue        http://www.last.fm/music/Blue   
469         475      Eminem      http://www.last.fm/music/Eminem   
498         504         HIM         http://www.last.fm/music/HIM   
640         646         RBD         http://www.last.fm/music/RBD   
695         701     Shakira     http://www.last.fm/music/Shakira   
1234       1243        Mika        http://www.last.fm/music/Mika   
1435       1444        Akon        http://www.last.fm/music/Akon   
1455       1464        ABBA        http://www.last.fm/music/ABBA   
4515       4609  Brainstorm  http://www.last.fm/music/Brainstorm   
4600       4704   Ladyhawke   http://www.last.fm/music/Ladyhawke   
4770       4880      Sandra      http://www.last.fm/music/Sandra   
4883       4993       Black       http://www.last.fm/music/Black   
6107       6232         BWO         http://www.last.fm/music/BWO   
14459     15171   Discovery   http://www.last.fm/music/Discovery   
16155     17084         OBK         http://www.last.fm/music/OBK   
17409     18460  Discobitch  http://www.last.fm/music/Discobitch   

                                              pictureURL  
65     http://userserve-ak.last.fm/serve/252/21564957...  
282    http://userserve-ak.last.fm/serve/252/53023109...  
314    http://userserve-ak.last.fm/serve/252/39646735...  
334    http://userserve-ak.last.fm/serve/252/49248521...  
456    http://userserve-ak.last.fm/serve/252/60025255...  
469    http://userserve-ak.last.fm/serve/252/29959269...  
498     http://userserve-ak.last.fm/serve/252/148293.jpg  
640    http://userserve-ak.last.fm/serve/252/59211131...  
695    http://userserve-ak.last.fm/serve/252/52116105...  
1234   http://userserve-ak.last.fm/serve/252/35401787...  
1435   http://userserve-ak.last.fm/serve/252/35787629...  
1455   http://userserve-ak.last.fm/serve/252/10441471...  
4515   http://userserve-ak.last.fm/serve/252/38523601...  
4600   http://userserve-ak.last.fm/serve/252/51100123...  
4770    http://userserve-ak.last.fm/serve/252/357981.jpg  
4883   http://userserve-ak.last.fm/serve/252/27521079...  
6107   http://userserve-ak.last.fm/serve/252/30550357...  
14459  http://userserve-ak.last.fm/serve/252/36091507...  
16155  http://userserve-ak.last.fm/serve/252/20392477...  
17409  http://userserve-ak.last.fm/serve/252/25413409...